aws/aws.go

```diff
@@ -0,0 +1,85 @@
+package aws
+
+import (
+	awsgo "github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/endpoints"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/gruntwork-io/gruntwork-cli/errors"
+)
+
+// Returns a list of all AWS regions
+func getAllRegions() []string {
+	resolver := endpoints.DefaultResolver()
+	partitions := resolver.(endpoints.EnumPartitions).Partitions()
+
+	var regions []string
+	for _, p := range partitions {
+		for id := range p.Regions() {
+			regions = append(regions, id)
+		}
+	}
+
+	return regions
+}
+
+// Returns a list of all resource identifiers in an AWS region from given account resources
+func getResourceIdenfiersForRegion(account *AwsAccountResources, resource string, region string) []string {
+	for _, res := range account.Resources[resource] {
+		if res.RegionName == region {
+			return res.ResourceIdentifiers
+		}
+	}
+
+	return make([]string, 0)
+}
+
+// GetAllResources - Lists all aws resources
+func GetAllResources() (*AwsAccountResources, error) {
+	account := AwsAccountResources{
+		Resources: make(map[string][]AwsRegionResources),
+	}
+
+	for _, region := range getAllRegions() {
+		session, err := session.NewSession(&awsgo.Config{
+			Region: awsgo.String(region)},
+		)
+
+		if err != nil {
+			return nil, errors.WithStackTrace(err)
+		}
+
+		instanceIds, err := getAllEc2Instances(session, region)
+		if err != nil {
+			return nil, errors.WithStackTrace(err)
+		}
+
+		regionResources := AwsRegionResources{
+			RegionName:          region,
+			ResourceIdentifiers: awsgo.StringValueSlice(instanceIds),
+		}
+
+		account.Resources["ec2"] = append(account.Resources["ec2"], regionResources)
```

Comment:
I'm not sure I like having a map with arbitrary resource strings in it. It won't be obvious to a maintainer later what kind of resources may be in the map and it's prone to typos ("EC2" vs "ec2"). 

I'd recommend using the type system here and having lists of specific resources types, such as `EC2Instance` and (in the future), `ASG`, `ELB`, etc.

Reply:
Ah, in fact, given your need to loop over all the resources and nuke them, it probably makes sense to do something like this:

```go
interface AwsResources {
  Nuke(session *aws.Session) error 
}

type AwsRegionResources struct {
	RegionName         string
	Resources          []AwsResources
}

type EC2Instances struct {
  InstanceIds []string
}

// Implement the AwsResources interface
func (instance EC2Instance) Nuke(session *aws.Session) error {
    // This is what's currently in the nukeAllEc2Instances method
}
```

Now `NukeAllResources` can just loop over all the resources:

```go
for _, region := range getAllRegions() {
  resourcesInRegion := account.Resources[region]
  for _, resources := resourcesInRegion.Resources {
    if err := resources.Nuke(session); err != nil {
      // ...
    }
  }
}
```